FLUID-5249: New documentation covering framework reorganisation under FLUID-5249

amb26 opened the pull request:
and other JIRAs. Documentation for new features for ContextAwareness (FLUID-5264) and Priority (FLUID-5506) features

colinbdclark commented on src/documents/APIChangesFrom1_5To2_0.md:
```diff
@@ -18,10 +18,9 @@ Rename "fluid.prefs.enactors" to "fluid.prefs.enactor"
 
 <div class="infusion-docs-note"><strong>Note:</strong> According to the [comment](https://github.com/fluid-project/infusion/blob/master/src/framework/core/js/FluidView.js#L38-L39) on the implementation for relay components, in Infusion 2.0, relay components will be renamed back to its original names. If the rename has been made, this section can be ignored.</div>
```
Can we update this comment to clearly describe the reality that the base component grades have been renamed?

Also, can we document other 1.5-2.0 changes such as the removal of demands blocks, non-dynamic invokers, etc.?

acheetham replied:
I also notice that the list of options supported by all components has changed, we should probably mention that.

acheetham replied:
I also see that grade linkages have been removed, we should mention that

acheetham replied:
onAttach has been removed

amb26 replied:
gradeLinkage to me counts as a "less widely-used feature" so I added an entry to it in the DeprecationsIn1_5 page referred to at the head of the list

acheetham replied:
thanks for the pointer, @amb26, I forgot to check the deprecations page.

acheetham replied:
Mention the change from fluid.event.makeEventFirer to fluid.makeEventFirer?

acheetham replied:
priority and namespace added to distributeOptions possibilities

colinbdclark commented on src/documents/ChangeApplierAPI.md:
```diff
@@ -446,29 +453,21 @@ These are not recommended for typical users of the framework, and are not stable
+Instantiating a ChangeApplier manually is no longer recommended in current versions of the framework. Its implementation is tightly bound into its location in 
+an IoC component tree and should be constructed by the IoC system itself.
 
 Under the old ChangeApplier semantics, the `model` handle was closed over and remained constant for the lifetime of the applier. The framework contained utilities such a `fluid.model.copyModel` which assisted the user in working with a model with changing contents whilst keeping its base reference constant. Under the new ChangeApplier semantics, the `model` base reference is not stable and in fac ...
```
Do we still need to retain this reference to the "old ChangeApplier semantics," since this is in fact now the "old old ChangeApplier semantic." In the spirit of 2.0 as a shedding of legacy concepts, can we also ditch this documentation or is it still crucial?

colinbdclark commented on src/documents/ChangeApplierAPI.md:
```diff
@@ -446,29 +453,21 @@ These are not recommended for typical users of the framework, and are not stable
 ### Operating transactions manually ###
 
-Similarly to manual construction of a ChangeApplier, this is not recommended for normal users of the framework. It may be appropriate for advanced users of the framework who are building higher-level frameworks layered over it.
+A user may be interested in economising on notifications to model updates - by batching these up into a single transaction, there will
```
Has this changed in Infusion master yet? If so, can we update the documentation?

acheetham replied:
The dash between "updates" and "by batching" should probably be a semicolon i.e. "... to model updates; by batching these..."

colinbdclark commented on src/documents/ComponentConfigurationOptions.md:
```diff
@@ -38,73 +36,72 @@ The following options are supported by all component grades:
+  </tr>
+  <tr>
+    <th>Notes</th>
+    <td>The Framework will create event firers for the listed events. It is the responsibility of the component to fire the events at the appropriate times.</td>
```
Is it worth clarifying here that the primary component lifecycle events, such as <code>onCreate</code> and <code>onDestroy</code> are fired automatically by the framework?

colinbdclark commented on src/documents/ComponentConfigurationOptions.md:
```diff
@@ -112,40 +109,59 @@ The following options are supported by all component grades:
-    <td><code>members</code> differ from <code>invokers</code> in that the arguments of members are not resolved at invocation time.
-        The right-hand-side may contain an <a href="ExpansionOfComponentOptions.md">expander</a> definition, which may perhaps itself resolve onto an <a href="Invokers.md">invoker</a>.</td>
+    <td>Both component developers and integrators can define listeners for events.
+<a href="Invokers.md">Invokers</a> and <a href="ExpansionOfComponentOptions.md">Expanders</a> can be used as listeners here. Note that as well as being a simple string holding the name of an event on this component, a listener key may also be a full <a href="IoCReferences.md">IoC Reference</a> to any other event held in the component tree (for example <code>"{parentComponent}.events.parentEvent"< ...
```
Is it true that an expander "can be used as a listener here," or is it slightly more subtle (i.e. the result of an expander can be used as a listener)?

colinbdclark replied:
typo "a **the** value"

colinbdclark replied:
My impression is that this note might will benefit from seeing specific example code (or links to pages where specific examples are available).

acheetham commented on src/documents/APIChangesFrom1_5To2_0.md:
```diff
@@ -48,14 +47,15 @@ fluid.default("fluid.parent", {
 ##### In 2.0 #####
 
-In Infusion 2.0 where relay components are introduced, sharing models no longer requires the change applier to be shared:
+In Infusion 2.0 where relay components are introduced, the change applier must not be configured separately - model sharing
```
It would be good if "change applier" here was a link to the ChangeApplier documentation. Also, perhaps instead of "must not" maybe "doesn't need to be"?

amb26 replied:
I think "must not be" - the user is pretty likely to corrupt this or future versions of the framework by attempting to directly configure any part of the model relay machinery. We do still support "changeApplierOptions" as a means of routing options to the applier - although currently we don't support any particularly useful options there.

acheetham commented on src/documents/ChangeApplierAPI.md:
```diff
@@ -8,9 +8,9 @@ This section explains and documents the various Javascript API calls for instant
+### Declarative style for listening to changes ###
 
-The declarative style for registering interest in change events uses an entry in the `modelListeners` options area of a `modelRelayComponent`. These listeners are attached to the applier during the construction process of the entire component (and its surrounding tree) and so will therefore become notified as part of the [initial transaction](ModelRelay.md#the-initial-transaction) - they will the ...
+The declarative style for registering interest in change events uses an entry in the `modelListeners` options area of a `modelComponent`. These listeners are attached to the applier during the construction process of the entire component (and its surrounding tree) and so will therefore become notified as part of the [initial transaction](ModelRelay.md#the-initial-transaction) - they will therefor ...
```
This might be a good opportunity to find a different word than "primordial" (which also occurs in the ModelRelay documentation page). Perhaps "initial"?

amb26 replied:
"initial" already signifies the same as it did in previous versions of the framework - that is, the "initial synchronised value of the model" taking into account all sources of initial values. I'm open to improved suggestions for "primordial" - perhaps "grandeval" or "sephiric"?

colinbdclark commented on src/documents/ComponentConfigurationOptions.md:
```diff
@@ -179,80 +195,51 @@ The following options are supported by all component grades:
   <tr>
     <th>Description</th>
-    <td>An object containing the data model to be used by the component.</td>
+    <td>An object providing instructions for how particular options should be merged when integrator options are merged with default values.</td>
```
In other parts of the documentation, we refer to "a record" rather than "an object." Is there a reason we don't do so here?

amb26 replied:
Perhaps because the pieces of documentation were written by different people at different times? : P - I've normalised this to "record" here

colinbdclark commented on src/documents/ComponentConfigurationOptions.md:
```diff
@@ -179,80 +195,51 @@ The following options are supported by all component grades:
   <tr>
     <th>Notes</th>
-    <td>If a <a href="ChangeApplier.md">ChangeApplier</a> is not provided using the <code><a href="#applier">applier</a></code> option, the Framework will create one for the provided model. </td>
+    <td>It is uncommon to need this option. The most common use case is to protect "exotic values" derived from some external library or framework from being corrupted by the options merging/expansion process by use of the "nomerge" policy.</td>
```
Is there an example of a kind of "exotic value" that one should typically protect? Or conversely, is there an example of a value that might seem like it needs be guarded, but doesn't? What about, say, a reference to a DOM object or other browser-implemented object (e.g. a Blob or an AudioContext or an Element)?

acheetham commented on src/documents/ChangeApplierAPI.md:
```diff
@@ -8,9 +8,9 @@ This section explains and documents the various Javascript API calls for instant
 
```
In the opening paragraph, it would be good if the first use of "ChangeApplier" was a link to the ChangeApplier page.

colinbdclark commented on src/documents/ComponentConfigurationOptions.md:
```diff
@@ -260,59 +247,62 @@ See also: [Component Grades](ComponentGrades.md)
-    <td>It is not necessary to provide an applier: By default, an applier will be created with <code>fluid.makeChangeApplier()</code>, using any options specified with <code><a href="#changeapplieroptions">changeApplierOptions</a></code>.
-
-This option is most commonly used to share a common ChangeApplier between components in a component tree: the <code>applier</code> option can be used to reference the ChangeApplier of another component in the tree.</td>
+    <td>Some special context names may be available within the subcomponent's definition block, for example <code>{source}</code> and <code>{sourcePath}</code> or <code>{arguments}</code>. <em>This framework facility will be replaced by a more declarative equivalent in time and should be used with caution.</em></td>
```
In what way should it be used with caution? Should we explicitly state that this is an unstable (or unsupported) API? Or should I apply some other form of caution (i.e. only using it for specific types of problems)?

amb26 replied:
I'm not really too sure what to say here. The API is not exactly "unstable" since we plan to support it for a while, and it has been supported for some time in the past. It is just one of these suspicious areas, like the old "init functions" etc. where we clearly haven't completely succeeded in understanding a problem area yet. Perhaps by "caution" I mean, "you should consult with someone before using this to confirm that it is appropriate".

colinbdclark replied:
I think I'd go ahead and say exactly this. "If you think you might need to use this feature, ask on < link to community forums > for advice" or something to that effect.

colinbdclark commented on src/documents/ComponentConfigurationOptions.md:
```diff
@@ -332,86 +321,76 @@ fluid.defaults("component.name", {
 </pre></td>
   </tr>
-  <tr>
+   <tr>
```
Extra whitespace.

colinbdclark commented on src/documents/ComponentConfigurationOptions.md:
```diff
@@ -332,86 +321,76 @@ fluid.defaults("component.name", {
-    <th>Notes</th>
-    <td>Both component developers and integrators can define listeners for events.
-<a href="Invokers.md">Invokers</a> and <a href="ExpansionOfComponentOptions.md">Expanders</a> can be used as listeners here. Note that as well as being a simple string holding the name of an event on this component, a listener key may also be a full <a href="IoCReferences.md">IoC Reference</a> to any other event held in the component tree (for example <code>"{parentComponent}.events.parentEvent"< ...
+    <td>A set of rules or constraints linking values held in this model to those elsewhere in the component tree (or to other values within this model)</td>
```
Is it worth noting here that these constraints are specific, and are operated by the Infusion model transformation, or is this too much detail for this page?

amb26 replied:
The link is below in the "See also" section

colinbdclark commented on src/documents/ComponentConfigurationOptions.md:
```diff
@@ -468,19 +446,51 @@ Component developers are free to define their own additional options.
 Component developers are free to define their own additional options.
 
-### selectorsToIgnore ###
+**NOTE**: The Infusion Renderer system will be rewritten completely for the Infusion 2.0 release - the use of the current renderer and component hierarchy is not recommended.
```
Again, can we be more specific, at least by saying "rewritten completely before the final release of Infusion 2.0?" It's this kind of issue that makes me increasingly wonder if we should indeed be more frequently updating our major revision numbers. The question of "when" is very fuzzy with our current plan to issue a long series of not-fully-roadmapped betas.

acheetham replied:
We have a "note" style that would set this off better visually, making it more noticable. This would be a good place to use it.

amb26 replied:
Can you refer me to an example of using this "note" style?

colinbdclark commented on src/documents/ComponentGrades.md:
```diff
@@ -14,121 +14,87 @@ The Infusion Framework already contains several predefined component grades that
-                <em>
-                    <strong>NOTE:</strong> for the Infusion 2.0 release this grade will become redundant as it will be the default for every grade
-                </em>
+                A "little" component is the most basic component: it supports options merging with defaults (<a href="tutorial-gettingStartedWithInfusion/BasicComponentCreation-LittleComponents.md">Little Components</a>), as well as
```
It's debatable, but I think I'd prefer italicization to quoting here, as in "A _little component_ is the most basic component..."

colinbdclark replied:
And should we even still be calling these "little" components now that they've been renamed? I'd recommend we ditch the prefix and rename the appropriate documentation pages.

acheetham replied:
This link to BasicComponentCreation-LittleComponents.md is broken, and the link text is in appropriate.

colinbdclark commented on src/documents/ComponentGrades.md:
```diff
@@ -14,121 +14,87 @@ The Infusion Framework already contains several predefined component grades that
-                    <strong>NOTE:</strong> for the Infusion 2.0 release this grade will become redundant as it will be the default for every grade
-                </em>
+                A "little" component is the most basic component: it supports options merging with defaults (<a href="tutorial-gettingStartedWithInfusion/BasicComponentCreation-LittleComponents.md">Little Components</a>), as well as
+                that instantiating event firers based on default framework events (onCreate, onDestroy, onDetach) 
```
Typo: unnecesary "that".

colinbdclark commented on src/documents/ComponentGrades.md:
```diff
@@ -14,121 +14,87 @@ The Infusion Framework already contains several predefined component grades that
+            <td><code>fluid.modelComponent</code></td>
             <td>
-                A "little" component is the most basic component: it supports options merging with defaults (<a href="tutorial-gettingStartedWithInfusion/BasicComponentCreation-LittleComponents.md">Little Components</a>). All Fluid components are derived from this grade, and in general all things not derived from this grade are non-components (e.g. plain functions, or model transformation transfo ...
+                A "model" component is a little component that additionally provides supports for a component's model, defined in the components options, and operations on it (<a href="tutorial-gettingStartedWithInfusion/ModelComponents.md">Tutorial - Model Components</a>).
```
Apostrophe missing: "components" -> "component's"

colinbdclark replied:
Is it worth referencing what these "operations" are? i.e. that the component a) has a ChangeApplier automatically assigned to it, and b) supports relays between models? Or is it better to keep it simple and general on this page?

colinbdclark commented on src/documents/ComponentGrades.md:
```diff
@@ -14,121 +14,87 @@ The Infusion Framework already contains several predefined component grades that
+            <td><code>fluid.viewComponent</code></td>
             <td>
-                A "model" component is already a little component that additionally provides supports for a component's model, defined in the components options, and operations on it (<a href="tutorial-gettingStartedWithInfusion/ModelComponents.md">Tutorial - Model Components</a>).
+                A "view" component is a fluid.modelComponent is bound to a DOM container node, holds a <a href="DOMBinder.md">DOM Binder</a> and supports a view (<a href="tutorial-gettingStartedWithInfusion/ViewComponents.md">Tutorial - View Components).
```
typo: missing "that." "A "view" component is a fluid.modelComponent is bound to a DOM..." -> "A "view" component is a fluid.modelComponent **that** is bound to a DOM..."

colinbdclark commented on src/documents/ComponentGrades.md:
```diff
@@ -14,121 +14,87 @@ The Infusion Framework already contains several predefined component grades that
-            <td>fluid.rendererRelayComponent</td>
-            <td>
-                A "renderer" component is already a vew component that bears a renderer. There are additional features provided by this component grade specified on the <a href="tutorial-gettingStartedWithInfusion/RendererComponents.md#useful-functions-and-events">Useful functions and events</a> section of the <a href="tutorial-gettingStartedWithInfusion/RendererComponents.md">Tutorial - Renderer ...
+                A "renderer" component is a view component that also bears a renderer. There are additional features provided by this component grade specified on the <a href="RendererComponents.md#useful-functions-and-events">Useful functions and events</a> section of the <a href="tutorial-gettingStartedWithInfusion/RendererComponents.md">Tutorial - Renderer Components</a> page
```
Does it "bear" a Renderer, or does it implement some kind of lifecycle for rendering? A renderer, currently, is somewhat different from, say, a ChangeApplier, isn't it?

colinbdclark commented on src/documents/ComponentGrades.md:
```diff
@@ -14,121 +14,87 @@ The Infusion Framework already contains several predefined component grades that
-A component's grades should be specified using the `gradeNames` option in the components defaults block, as shown in the examples below. The `gradeNames` option holds a string or Array of Strings.
-
-<div class="infusion-docs-note"><strong>Note:</strong> In the examples below, the <code>autoInit</code> flag is not actually a grade, but is added to the <code>gradeNames</code> array to control how the component is created. See <a href="#initializing-graded-components">Initializing Graded Components</a> below for more information about the <code>autoInit</code> flag. The <code>autoInit</code> fl ...
+A component's grades should be specified using the `gradeNames` option in the components defaults block, as shown in the examples below. The `gradeNames` option holds a `String` or `Array of String`.
```
What happens if <code>gradeNames</code> is omitted from a component's definition?

colinbdclark commented on src/documents/ComponentGrades.md:
```diff
@@ -14,121 +14,87 @@ The Infusion Framework already contains several predefined component grades that
 
-Since the `fluid.defaults` directive introduces a grade into the system, various components can be composed to create new ones. Options, fields and methods introduced by the ancestor grades will be merged. The merging happens, firstly in hierarchical order (grades comprising the ancestor grade are resolved before the actual component grades resolution) and secondly in the left-to-right order (def ...
+Since the `fluid.defaults` directive introduces a grade into the system, various components can be composed to create new ones. Options, fields and methods introduced by the ancestor grades will be merged. 
+The merging happens, firstly in hierarchical order (grades comprising the ancestor grade are resolved before the actual component grades resolution) and secondly in the right-to-left order (defaults from the grade on the left taking precedence over the defaults from the grade on the right, more details can be found at the JIRA [FLUID-5085](http://issues.fluidproject.org/browse/FLUID-5085)). For e ...
```
Rather than pointing the reader to a rather context-specific JIRA ticket, can we just describe those "more details" here or elsewhere in the documentation?

amb26 replied:
The details are pretty esoteric and not really of interest to even moderately regular users of the framework - given in practice we find that "the expected thing mostly happens" when people write grade lists. The useful pointer in the JIRA is that, in particular, we do not operate the standard "C3 linearization algorithm" (or anything like it) that is common in OO-like systems - https://en.wikipedia.org/wiki/C3_linearization - I don't know if it's worth trying to make this point or highlight it anywhere else/otherwise.

colinbdclark replied:
I'd suggest ditching the reference to the JIRA, and adding a reference that is essentially the same as what you say here: (roughly) "This is in contrast to the C3 linearization algorithm (link) that is common in other object oriented inheritance schemes."

colinbdclark commented on src/documents/ContextAwareness.md:
```diff
@@ -0,0 +1,327 @@
+category: API
+---
+
+The ContextAwareness API of Infusion provides a powerful suite of grades and utilities that allow a component to be responsive to aspects of its contexts in a flexible and open-ended way.
```
I think we can use "context" singular rather than "contexts" here, though I'm open to debate.

acheetham replied:
The word context should be a link to the Contexts documentation.

colinbdclark commented on src/documents/ContextAwareness.md:
```diff
@@ -0,0 +1,327 @@
+A `contextAware` component can be responsive to contexts, whose influence is organised as a set of mutually orthogonal "adaptations", where the set of adaptations itself is open to extension by 
+integrators and adopters. The primary grade enabling a component to opt into this scheme is named `fluid.contextAware`. Any component derived from this grade will advertise an
+area of its options named `contextAwareness` which organises the rules for its adaptation in a hierarchical way - at the top level by "adaptation" and then at the nested level by
+"checks".
```
I think it would be hugely helpful, by this point in the documentation, to provide a concrete example of a kind of "context awareness" that a component may want to provide. Perhaps a common kind of feature or browser detection, for example? Readers will benefit from being able to map the terminology to a specific example while they're reading.

colinbdclark commented on src/documents/ContextAwareness.md:
```diff
@@ -0,0 +1,327 @@
+<table>
+    <thead>
+        <tr>
+            <th colspan="3">Members of a <code>adaptationRecord</code> entry within the <code>contextAwareness</code> block of a <code>fluid.contextAware</code> component</th>
```
Typo: "Members of a adaptationRecord" -> "Members of **an** adaptationRecord"

acheetham replied:
It would be very helpful to the reader if there was a code example before this table. It would help to clarify the relationship between adaptation records, contextAwareness blocks, etc. much more easily than linear text.

colinbdclark commented on src/documents/ContextAwareness.md:
```diff
@@ -0,0 +1,327 @@
+
+### Example `contextAwareness` record
+
+The most adaptible component in the framework is currently the [Uploader](togo/Uploader.md) which currently can respond to three "dimensions" of adaptation. Two of these,
```
I think it might be better to use an example that isn't our own component, but nonetheless is a simple example of something related to browser feature detection.

colinbdclark commented on src/documents/ContextAwareness.md:
````diff
@@ -0,0 +1,327 @@
+});
+```
+
+`technology` refers to the implementation technology of the uploader. For previous releases, this supported a Flash-based engine as well as an HTML engine targetted at a back-levelled HTML
````
Too much historical detail clouds the clarity of this example.

colinbdclark commented on src/documents/ContextAwareness.md:
```diff
@@ -0,0 +1,327 @@
+
+Note that the combined effect of the first two defaults blocks shown in this example can be achieved "all-in-one" by a single call to the dedicated utility `fluid.contextAware.makeAdaptation`. 
+
+## Making contexts visible and removing them with `fluid.contextAware.makeChecks` and `fluid.contextAware.forgetChecks`
```
When would I want to make checks visible and removing them? For test cases?

acheetham replied:
This isn't a docs comment, but actually an API comment: If I'm understanding correctly, these functions are actually making and forgetting *contexts* that could be checked with a checkRecord. If this is correct, then "makeCheck" doesn't seem like the right name - should it be "makeContext" ? If my understanding isn't correct, then this needs to be clarified in this documentation.

colinbdclark commented on src/documents/Contexts.md:
```diff
@@ -61,15 +52,106 @@ at a particular component location in the tree are the same, as described in the
+2. Arrange for there to be a single instance of a component
+
+The Infusion framework supplies two special grades, which allow the designer to arrange for both of these
+points to be controlled separately. For a start, any component which is supplied the grade `fluid.resolveRoot`,
```
It's unclear why this grade is called <code>fluid.resolveRoot</code>. Based on the documentation, wouldn't it be more descriptive to call it something like <code>fluid.globallyVisible</code>? And below, instead of <code>fluid.resolveRootSingle</code> just <code>fluid.singleton</code>? There may be a reason for the current naming, but it isn't clear from the documentation at the moment.

acheetham replied:
I agree with Colin's suggestions on this, they're much more clear.

amb26 replied:
I'd like the two names that we pick to have a clear verbal relationship to one another - since the grade currently named "resolveRootSingle" inherits from the one named "resolveRoot"

colinbdclark commented on src/documents/Enactors.md:
````diff
@@ -25,16 +25,16 @@ Each of these is explained below.
 });
 ```
 
 Enactors are, by default, [standard relay components](ComponentGrades.md
````
Lingering reference to "standard relay components." Should be "model components."

colinbdclark commented on src/documents/ExpansionOfComponentOptions.md:
```diff
@@ -52,6 +52,8 @@ The basic form of an expander record is very similar to that of an [Invoker](Inv
     </tbody>
 </table>
 
+Note that unlike an invoker which evalutes its arguments every time it is used by its caller, an expander evaluates only once - when the component whose options hold it is constructing.
```
Typo: "evalutes" -> "evaluates"

colinbdclark commented on src/documents/FrameworkConcepts.md:
```diff
@@ -8,31 +8,56 @@ The Fluid Infusion Framework is built on a number of important concepts. This pa
+The greatest goal of the framework is *transparency* &#8212; data is pure data, and markup is pure markup. Infusion doesn't impose any inheritance hierarchies or strict contracts on how you structure your data model.
 
-In Infusion, a tree of objects has the status of being a Model Object if it can successfully be copied by the framework utility function `fluid.copy()`. This function duplicates an object tree in a very clear way: each object is cloned recursively, except for immutable values such as strings, numbers, functions, etc. which are copied across. If everything that is important to you about the object ...
+In Infusion, a tree of objects has the status of being a Model Object if it can successfully be copied by the framework utility function `fluid.copy()`. 
```
"Model Objects" -> "model"

colinbdclark commented on src/documents/FrameworkConcepts.md:
```diff
@@ -8,31 +8,56 @@ The Fluid Infusion Framework is built on a number of important concepts. This pa
+consist of only the following types: `Array`, `Object`, `String`, `Number` and `Boolean`. In addition, the following values for `Number` should not be used - `Infinity`, `-Infinity` and `NaN`. The value
+`undefined` also can not be safely stored within a model.
 
 An analogy from another environment, Java, is a [POJO](http://en.wikipedia.org/wiki/POJO) or "Plain Old Java Object". The J could just as easily be substituted for "JavaScript" in this case.
```
Is this analogy still useful? I imagine Java programmers still think of POJOs as more "object-oriented" than what an Infusion model is (e.g. POJOs still may well have getters and setters)

colinbdclark commented on src/documents/FrameworkConcepts.md:
```diff
@@ -8,31 +8,56 @@ The Fluid Infusion Framework is built on a number of important concepts. This pa
 ## EL Paths
 
-In some parts of the framework, we refer to "EL expressions". This is a somewhat historical phrase that perhaps sounds like it says more than it is trying to. All we mean by EL expressions are dot-separated paths built of names—for example if you had defined an object `zar` with a member `boo` which has a member `baz`, you could access the nested JavaScript property by writing the expression `zar ...
+In some parts of the framework, we refer to "EL expressions". This is a somewhat historical phrase that perhaps sounds like it says more than it is trying to. 
```
Given that we support IoC expressions, is it true that our term of "EL expressions" is historical and saying more than it tries to? Or can this comment be removed?

amb26 replied:
I don't know - what do you think? IoC expressions are not identical with the set of EL expressions - you could say that "An IoC expression consists of a context expression followed by an EL expression".

colinbdclark commented on src/documents/FrameworkConcepts.md:
```diff
@@ -8,31 +8,56 @@ The Fluid Infusion Framework is built on a number of important concepts. This pa
+you could access the nested JavaScript property by writing the expression `zar.boo.baz`. 
+If it were held in a string value, this would become an EL expression &#8212; that is, the string `"zar.boo.baz"` is an EL expression which designates the same piece of data we just referenced. 
+The framework includes machinery for interpreting such expressions held in strings rather than at the JavaScript language level. 
+This is useful because it abstracts references to pieces of data from the actual data itself - and to even store these references in documents separately from a running program. 
```
The second half of this sentence "to even store these references in documents" seems ungrammatical and might require some clarification. It's a very profound point, but unclear from this fragment.

colinbdclark commented on src/documents/InfusionEventSystem.md:
```diff
@@ -46,11 +46,13 @@ A top-level options block named **`events`** is supported on every component der
   </tr>
 </table>
 
+**NOTE**: _`preventable` events are very rarely used and will soon be deprecated in the framework_.
```
When?

amb26 replied:
I guess I can't firmly predict that they will indeed be deprecated - although I'm not particularly looking forward to a community discussion on whether they should be deprecated or not. I added this sentence because I noticed that in the first draft of his Infusionised components, @the-t-in-rtf had made each of his events "preventable" by default and I wanted to highlight in the docs that this represented at the very least an unusual implementation choice. I used to think that there was no valid use case for an event which terminated processing, since we had no detailed control over the positioning of event listeners in a chain - but with the introduction of priority constraints, this is no longer true. You could now consider a "preventable" event is one that operates a semantic similar to the kinds of "transforming promise chain pseudoevents" that are starting to appear, where a rejection is received in the middle of the chain. I think that with the introduction of promise idioms into the framework, we will start to have other kinds of events built into the core... such as "enlatening events" to handle the use case that @the-t-in-rtf  ran into with his pouchdb wrapper, where he has a piece of on-destruction cleanup that he needs to perform that is asynchronous - and that we should not advertise that the component is destroyed until it is done (that is, "afterDestroy" could be delayed to an arbitrary point following "onDestroy" rather than occuring synchronously after it). 

I could either say "may soon be deprecated" or as we have in other places, "talk to us if you want to try to use these".

colinbdclark commented on src/documents/InfusionEventSystem.md:
```diff
@@ -141,26 +203,19 @@ Traditional procedural APIs corresponding to all the above declarations exist. H
 The Fluid event system is operated by instances of an *event firer* which are created by a call to `fluid.event.makeEventFirer()`. It is recommended that users don't construct event firers by hand but instead rely on the framework's facilities for automatically constructing these given event blocks in [component options](ComponentConfigurationOptions.md). The signature of `fluid.event.makeEventFi ...
```
Has this signature changed? If not, when will it? Should we remove the documentation for <code>makeEventFirer</code>?

amb26 replied:
It did change, yes - together with the signature for addListener (in its later arguments)

colinbdclark commented on src/documents/IoCReferences.md:
```diff
@@ -169,13 +183,13 @@ fluid.defaults("fluid.prefs.enactor.tableOfContents", {
 The example below includes several IoC references. All of them are inside a subcomponent declaration and all include `{controllers}`, which in this case is a reference to the parent component. Specifically:
 
-* `{controllers}.model` and `{controllers}.applier` are references to the model and applier that are members of the parent component;
+* `{controllers}.model` is references to the model that is a members of the parent component - note that this reference sets up a permanent [model relay](ModelRelay.md) between these two models;
```
There's a grammatical issue that has me confused about what this sentence is saying.

acheetham replied:
change "is references" to "is a reference" and "is a members" to "is a member"

colinbdclark commented on src/documents/IoCSS.md:
```diff
@@ -47,74 +52,78 @@ fluid.uiOptions(".my-uio-container", {
+In the `distributeOptions` block above, the context `{that templateLoader}` is an IoCSS expression which designates one or more of the child components of UI Options that are to receive the user's option. The syntax and meaning of these expressions is defined below.
+
+As well as converting the exposed options structure of a component into a more compact form, `distributeOptions` is also a powerful tool for maintaining API stability for a component or family of components. 
+Since the binding of IoCSS selectors such as `that templateLoader` onto child components is flexible, the component tree could be refactored in quite an aggressive way without requiring changes in either the user's configuration, or even the `distributeOptions` block itself. 
```
This is such an awesome point! \o/

colinbdclark commented on src/documents/IoCSS.md:
````diff
@@ -47,74 +52,78 @@ fluid.uiOptions(".my-uio-container", {
 ```
 
-## Example: record ##
+The `sessionManager`, wherever it is in the set of child components, will be granted the additional grade `kettle.sessionManager.io`.
````
Given the refactoring of Kettle, is the example somewhat misleading now? Should we use a different example?

amb26 replied:
I don't know yet, I haven't completed the reimplementation of session-secured WebSockets support : P

colinbdclark commented on src/documents/IoCSS.md:
````diff
@@ -140,3 +149,61 @@ fluid.defaults("fluid.moduleLayoutHandler", {
+        target: "{/ fluid.viewComponent}.options.gradeNames" // <- use of IoCSS root context "/"
+    }
+});
+```
````
The variety and simplicity of all these examples is great. Bonus points for sneaking in the Star Trek reference. I love that episode of TNG.

colinbdclark commented on src/documents/IoCTestingFramework.md:
```diff
@@ -11,18 +11,18 @@ As well as creating an idiomatic way of writing *integration tests* addressed at
 The concept of *context* in Infusion IoC is derived from the entire collection of components held in an IoC component tree.
-The behaviour of each component is potentially altered by all of the other components which are *in scope* from the site of the component
-under consideration - for a detailed guide to the operation of scope within Infusion IoC, please consult the page on [Contexts](Contexts.md).
+The behaviour of each component is potentially altered by all of the other components with which is deployed 
```
Grammatical issue here. "with which **it** is deployed"?

colinbdclark commented on src/documents/IoCTestingFramework.md:
```diff
@@ -52,8 +52,8 @@ including a summary of work in progress on [December 5th](http://lists.idrc.ocad
 
-The IoC Testing framework requires the use of two new kinds of Fluid component, which are packaged as ["grades"](ComponentGrades.md) within the implementation in the
-file IoCTestUtils.js. In order to make use of the framework, the tester must derive their own component types from these grades, and assemble
+Writing fixtures using the IoC Testing framework requires the test implementor to derive from two special ["grades"](ComponentGrades.md) which are packaged within the testing framework implementation in the
```
typo: "implementor" -> "implementer"

amb26 replied:
permitted alternative, I think

acheetham commented on src/documents/Priorities.md:
```diff
@@ -0,0 +1,107 @@
+
+The priority system is designed to be "open" in the face of evolving application designs that are worked on by different groups. As well as standard
+fixed priorities (represented by numbers or _extremal values_ such as `"first"` or `"last"`), Infusion priorities may be _constraint-based_ in which
+an element defines its priority only with respect to another element which is identified by _namespace_. These constraint-based priorities are
```
Either here or in the table below, the word "namespace" should be explained, or be a link to a page that explains what we mean.

colinbdclark commented on src/documents/IoCTestingFramework.md:
```diff
@@ -70,12 +70,14 @@ This simple example shows the testing of a simple component, `fluid.tests.cat` w
-    };
+
+fluid.tests.cat.makeSound = function () {
+    return "meow";
```
https://www.youtube.com/watch?v=kxRX6LXDpWs

amb26 replied:
https://www.youtube.com/watch?v=8Caa-QEY-FA :)

acheetham commented on src/documents/Priorities.md:
```diff
@@ -0,0 +1,107 @@
+listener, or a set of options or grades may take priority over another during options merging.
+
+The priority system is designed to be "open" in the face of evolving application designs that are worked on by different groups. As well as standard
+fixed priorities (represented by numbers or _extremal values_ such as `"first"` or `"last"`), Infusion priorities may be _constraint-based_ in which
```
The word "extremal" is a bit unfamiliar. Perhaps "boundary"? "edge"?

acheetham commented on src/documents/Priorities.md:
```diff
@@ -0,0 +1,107 @@
+unless a further constraint positions a third listener in between them.</td><td><code>"before:bindMarkup"</code>, <code>"after:computeLayout"</code></td></tr>
+
+<tr><td>Extremal Priority</td>
+<td>Either of the strings <code>first</code> or <code>last</code>, either by themselves, or followed by an <strong>extremal priority class</strong> separated by a colon. Elements annotated <code>first</code> or
```
The two possible values for an "extremal priority class" should be more clearly set off and stated earlier. Right now, they're buried at the end of a long narrative, leaving the reader wondering "What is this class? Where is it defined? Do I make it up, or does the framework offer it?" etc...

acheetham commented on src/documents/Priorities.md:
```diff
@@ -0,0 +1,107 @@
+---
+title: Priorities
+layout: default
+category: API
```
I'm not convinced this page should be in the API category. It also needs to be added to the Table of Contents, in whichever category we end up putting it in.

acheetham commented on src/documents/Priorities.md:
```diff
@@ -0,0 +1,107 @@
+resulting from grade merging. For options distributions, the natural ordering will be derived instead from the tree's topology (distributions which travelled a longer distance will
+be weaker than those which travelled a shorter distance), and numeric priorities should not be used.
+
+## Example of priority resolution for listeners
```
I would suggest putting this example *after* the "Table of supported sites for priorities"

amb26 replied:
I would like to invite my reviewers to normalise this issue amongst themselves - earlier in the ContextAwareness.md docs, @colinbdclark  suggested putting an example before the main detailed explanation. Should we settle on a standard layout with example both before and afterwards, or just in one place?

acheetham commented on src/documents/Priorities.md:
```diff
@@ -0,0 +1,107 @@
+  <tr>
+  <td><a href="InfusionEventSystem.md#registering-a-listener-to-an-event"><code>listeners</code></a></td>
+  <td>Lower numeric numbers, <code>last</code>, and <code>after</code> will fire later than others</td>
+  <td>Full support for constraints, uniquifying based on <code>namespace</code></td>
```
What does "uniquifying" mean? If it means "adding a namespace after the before/after", then the 'no support for uniquifying' phrases below makes= no sense, because above, constraints are described as including these namespaces i.e. there is no indication that you can have before/after *without* a namespace. Regardless, please try to find some other way to phrase this besides "uniquifying".

acheetham commented on src/documents/Priorities.md:
```diff
@@ -0,0 +1,107 @@
+  </tr>
+  <tr><td><a href="SubcomponentDeclaration.md"><code>components</code></a></td>
+  <td>Ordering as for <code>listeners</code>. Components with lower priority will be constructed later.</td>
+  <td>Note that there is <strong>no</strong> support for <code>namespace</code> for component declarations. The use of priorities for subcomponent declarations is <strong>not recommended</strong>. You should always be able 
```
How does "no support for namespace" differ from "no uniquifying based on namespace"? And does it mean no support for constraints at all? The others refer to constraints directly, so this one should, too, one way or the other.

amb26 replied:
No support for namespaces implies no support for constraints, yes - since constraints use the namespace names for references. I've adjusted this section a little to account for the fact that it describes both support for namespaces in themselves as well as the priorities that might be derived from them.

acheetham commented on src/documents/Contexts.md:
```diff
@@ -37,15 +37,6 @@ not available for components at the head of the tree. It is also the most unstab
 it depends on particular naming of members in the tree rather than remaining stable if the tree is
 reorganised. However, it is the only route for matching a particular instance of a component rather
 than matching components by their type or name in general.
-1. The component's "nickname" - which by default is derived from the last path segment of its fully
```
[This note is not specific to this line]
Somewhere in this document, there needs to be a link to the ContextAwareness API page.

acheetham commented on src/documents/ContextAwareness.md:
```diff
@@ -0,0 +1,327 @@
+---
+
+The ContextAwareness API of Infusion provides a powerful suite of grades and utilities that allow a component to be responsive to aspects of its contexts in a flexible and open-ended way.
+For lightweight cases of context adaptation, the [`distributeOptions`](IoCSS.md) scheme suffices, in that it enables a component to be responsive to broadcasts "advising" it from around the
```
This opening paragraph is too long and too all-encompassing. Our stylesheet renders the first paragraph in larger font, making it a quick at-a-glance into to what this page is about. I suggest ending the opening paragraph for this page after that first sentence.
The rest of this paragraph uses words and phrases that are not explained, and so are confusing to the reader (e.g. "adaptation", "two or more "axes" or "dimensions" of adaptation"). As Colin suggests below, an example of what the heck this is all about might be helpful before going into details.

acheetham replied:
I'm not convinced that "broadcast" is the right word here (and in the rest of this page). It's possible that I'm not actually understanding, but "broadcast" implies sending information out, for anyone to hear, not knowing or caring who listens or acts on the information – more like events. But if I'm understanding this API correctly (which I may not be!), this sounds much more targeted and deliberate than that, in which case I would encourage you to come up with better terminology than "broadcast".

amb26 replied:
I think "broadcast" fits because there may be zero or more components which could receive the broadcast.

acheetham commented on src/documents/ContextAwareness.md:
```diff
@@ -0,0 +1,327 @@
+The ContextAwareness API of Infusion provides a powerful suite of grades and utilities that allow a component to be responsive to aspects of its contexts in a flexible and open-ended way.
+For lightweight cases of context adaptation, the [`distributeOptions`](IoCSS.md) scheme suffices, in that it enables a component to be responsive to broadcasts "advising" it from around the
+component tree in a free-form and controllable away. However, when a component needs to take a clearer role in responding to potentially many kinds of requests for adaptation, where these
+can be organised around two or more "axes" or "dimensions" of adaptation that can be clearly named and identified, it should opt into this more heavywight scheme by implementing the 
```
Instead of "it should opt into this more heavyweight scheme by implementing the 'fluid.contextAware' grade" perhaps just say "it should use the 'fluid.contextAware' grade"

acheetham commented on src/documents/ContextAwareness.md:
```diff
@@ -0,0 +1,327 @@
+---
+title: ContextAwareness API
+layout: default
+category: API
```
We need to make sure this page appears in the Table of Contents for this category.

colinbdclark commented on src/documents/ModelRelay.md:
```diff
@@ -4,33 +4,58 @@ layout: default
 ### The Initial Transaction ###
 
-Whenever a new model-bearing component (or an entire tree of model-bearing components) constructs, there will be a particular large style of update transaction known as an __initial transaction__.   This is extremely similar to any other synchronisation transaction caused by a model relay update, although it will typically involve more data since all of the initial values of all the involved mode ...
+Whenever a new model-bearing component (or an entire tree of model-bearing components) constructs, there will be a particular large style of update transaction known as an __initial transaction__.  
```
Typo: "particular large" -> "particularly large"?

amb26 replied:
I had meant "particular" - do you think it would be clearer if expressed as "a particular large update transaction" or could be rewritten completely?

colinbdclark replied:
Then are we missing a comma? "a particular, large update transaction?" There something mildly ambiguous about it as-is.

colinbdclark commented on src/documents/ModelRelay.md:
```diff
@@ -4,33 +4,58 @@ layout: default
+Whenever a new model-bearing component (or an entire tree of model-bearing components) constructs, there will be a particular large style of update transaction known as an __initial transaction__.  
+This is very similar to any other synchronisation transaction caused by a model relay update, although it will typically involve more data since all of the initial values of all the involved models must be taken into account &#8212; 
+these result from any of the normal sources for component configuration, including [defaults](https://github.com/fluid-project/infusion/blob/infusion-1.5/src/framework/core/js/Fluid.js#L1519-L1539), user-supplied values, 
+[distributed](IoCSS.md) options, etc.. During the initial transaction, any declaratively registered listeners will observe all of the new models go through the transition from holding their primordial value of `undefined` 
```
Stray period: "etc.."

amb26 replied:
Yes, you are correct - previously I had considered that an extra one was required.... http://english.stackexchange.com/questions/8382/when-etc-is-at-the-end-of-a-phrase-do-you-place-a-period-after-it

amb26 replied:
Probably stemming from my TeX background, where different spacing rules are applied for "end of sentence" periods and "end of abbreviation" periods : P

acheetham commented on src/documents/ContextAwareness.md:
```diff
@@ -0,0 +1,327 @@
+`fluid.contextAware` grade and perhaps also advertising some of its dimensions of adaptations in its own options. These dimensions themselves can always be extended by further contributions
+to the component's options through all the usual sources - direct options, subcomponent options, and options distributions from elsewhere in the tree.
+
+A `contextAware` component can be responsive to contexts, whose influence is organised as a set of mutually orthogonal "adaptations", where the set of adaptations itself is open to extension by 
```
Much of this paragraph repeats things already stated.

acheetham commented on src/documents/ContextAwareness.md:
```diff
@@ -0,0 +1,327 @@
+area of its options named `contextAwareness` which organises the rules for its adaptation in a hierarchical way - at the top level by "adaptation" and then at the nested level by
+"checks".
+
+## Structure of members in a `adaptationRecord`
```
The string "adaptationRecord" doesn't seem to appear anywhere in the code. If I'm right about this, then it should not be appearing the docs in this form, i.e. code-styled and camelCase. It is a conceptual thing, and so should be referred to as simply "adaptation record".

amb26 replied:
I think in future we will want to structure our "schematic" documentation around "named types of things" which can occur in various particular places in configuration - and also be automatically checked by our authoring tools. As such, we want "adaptationRecord" to be a real thing rather than merely a conceptual thing, and this represents the start of it...

acheetham commented on src/documents/ContextAwareness.md:
```diff
@@ -0,0 +1,327 @@
+
+## Structure of members in a `adaptationRecord`
+
+The top level structure of the `contextAwareness` record consists of a free hash of `adaptationName` strings to `adaptationRecord` structures. The `adaptationName` strings will be considered as
```
This sentence refers to "the 'contextAwareness' record" record as though it's something that the reader already knows about, which is not the case. It should be properly introduced.

acheetham commented on src/documents/ContextAwareness.md:
```diff
@@ -0,0 +1,327 @@
+## Structure of members in a `adaptationRecord`
+
+The top level structure of the `contextAwareness` record consists of a free hash of `adaptationName` strings to `adaptationRecord` structures. The `adaptationName` strings will be considered as
+namespaces for the purposes of priority resolution (see the <code>priority</code> entry in the following table). The elements of the `adaptationRecord` are
```
This reference to priority resolution should be a link to the Priorities documentation.

acheetham commented on src/documents/ContextAwareness.md:
```diff
@@ -0,0 +1,327 @@
+            <th colspan="3">Members of a <code>adaptationRecord</code> entry within the <code>contextAwareness</code> block of a <code>fluid.contextAware</code> component</th>
+        </tr>
+        <tr>
+            <th>Member</th>
```
Is 'Member' the 'adaptationName' referred to above? If so, the column header should probably be 'adaptationName'

acheetham commented on src/documents/ContextAwareness.md:
```diff
@@ -0,0 +1,327 @@
+    </tbody>
+</table>
+
+The result of the `contextAwareness` record is that a number of the elements within `checks` will be evaluated in the visible context, and result in a number of `gradeNames` which will then be
```
This information – the fact that this is about automatically adding grades to a component based on examinations of the surrounding context – needs to be presented much earlier in the document, right at the top. I'm gathering that this addition of grades is what you mean when you say "adaptation". That needs to be explained the very first time you use the word 'adaptation'.

amb26 replied:
Now part of intro

acheetham commented on src/documents/ContextAwareness.md:
```diff
@@ -0,0 +1,327 @@
+
+##Structure of members in a `checkRecord`
+
+The `checkRecord` structure which is used in the first row of this table is described now:
```
This should refer to "the table above" and not "this table".

acheetham commented on src/documents/ContextAwareness.md:
```diff
@@ -0,0 +1,327 @@
+---
+title: ContextAwareness API
```
What is the relationship of this API to the Progressive Enhancement stuff? They seem similar. Is this effectively a replacement for that?

amb26 replied:
Yes sorry, ProgressiveEnhancement is effectively abolished in this release in favour of ContextAwareness. I neglected to delete the docs page referring to it

acheetham commented on src/documents/ContextAwareness.md:
```diff
@@ -0,0 +1,327 @@
+        <tr>
+            <td><code>defaultGradeNames</code> (optional)</td>
+            <td><code>String</code> or <code>Array of String</code></td>
+            <td>One or more <code>gradeNames</code> to be the result if none of the <code>checks</code> entries matches</td>
```
If 'checks' is optional, what happens if none are specified? If this gradeName would, in that case, always be added, why put it in a contextAwareness block and not just directly in the gradeNames option? This should be explained.

amb26 replied:
I'm not sure how I could make this much clearer. I could talk about a possible process where one user had contributed a dimension where all they were sure about was the default grade in case no check matched, and further users then contributed checks which created conditions in which the default no longer always matched, but my feeling is that this would just represent a distraction to the user trying to learn from the docs. What do you think?

acheetham commented on src/documents/ContextAwareness.md:
```diff
@@ -0,0 +1,327 @@
+        <tr>
+            <td><code>checks</code> (optional)</td>
+            <td>Hash of <code>checkNamespace</code> to <code>checkRecord</code></td>
+            <td>A free hash of namespace names for checks, to entries describing how a context check is to be made, and what <code>gradeNames</code> should result if the check is successful.</td>
```
This would benefit from the addition of "(See 'Structure of members in a checkRecord' below for details.)" with a hyperlink in it.

acheetham commented on src/documents/ContextAwareness.md:
```diff
@@ -0,0 +1,327 @@
+        <tr>
+            <td><code>gradeNames</code> (optional)</td>
+            <td><code>String</code> or <code>Array of String</code></td>
+            <td>One or more <code>gradeNames</code> that will be returned out to the <code>contextAwareness</code> system if this check passes. The check will pass if the context value referenced in <code>contextValue</code> can be found, and
```
The conditions for a check passing should probably not be buried in the description of an optional member that isn't part of the determination. This might be better places in the 'equals' description.

acheetham replied:
This entry is optional; if no gradeNames are provided and the check passes, what happens?

amb26 replied:
In that case, no gradeNames are contributed - but on the other hand the defaultGradeNames are not contributed either. I think this is as would be expected?

acheetham commented on src/documents/ContextAwareness.md:
```diff
@@ -0,0 +1,327 @@
+        <tr>
+            <td><code>priority</code></td>
+            <td><code>Priority</code> value - see <a href="Priorities.md">Priorities</a> for a full explanation</td>
+            <td>The priority (if any) that the <code>gradeNames</code> resulting from this dimension should have over those resulting from any other dimension. This should be an entry of the form
```
The phrase "(if any)" implies that this member is optional, but the Member column doesn't say it is?

acheetham commented on src/documents/ContextAwareness.md:
````diff
@@ -0,0 +1,327 @@
+});
+```
+
+The defaults block `fluid.uploader.compatibility.distributor.1_3` contains an options distribution which appeals to the existence of a third dimension in the uploader's `contextAwareness`, named
````
"appeals to the existence of" is an unclear phrase. What does this mean? Does it *create* 'apiCompatibility'? If so, then just say so (if not, where does apiCompatibility come from?)

acheetham commented on src/documents/ContextAwareness.md:
```diff
@@ -0,0 +1,327 @@
+
+The defaults block `fluid.uploader.compatibility.distributor.1_3` contains an options distribution which appeals to the existence of a third dimension in the uploader's `contextAwareness`, named
+`apiCompatibility` &#8212; this can be done simply by arranging to broadcast the appropriate options into it. Once we have defined this options distribution, we actually need to construct a component
+a component instance which holds and operates them &#8212; this is done via the `fluid.constructSingle` line. This utility automatically arranges for a singleton instance, uniquified at the 
```
There's an extra "a component" here.

acheetham commented on src/documents/ContextAwareness.md:
````diff
@@ -0,0 +1,327 @@
+```
+
+The defaults block `fluid.uploader.compatibility.distributor.1_3` contains an options distribution which appeals to the existence of a third dimension in the uploader's `contextAwareness`, named
+`apiCompatibility` &#8212; this can be done simply by arranging to broadcast the appropriate options into it. Once we have defined this options distribution, we actually need to construct a component
````
Is this example of constructing a singleton instance a part of working with the contextAwareness api? Is all use of this api going to involve this? If not, then I think the example is introducing some unrelated and thus confusing things that cloud the issue. It would be better to have a more directly relevant example.

amb26 replied:
The process of constructing such a singleton is part of a usage of one style of the contextAwareness API - as it is subsumed into the fluid.contextAware.makeAdaptation convenience API. I'll add some more signposting material to this at the head as you suggest.

acheetham commented on src/documents/ContextAwareness.md:
```diff
@@ -0,0 +1,327 @@
+component tree's top level by the type `singleRootType` which has a very similar function to the option of the same name consumed by the `fluid.resolveRootSingle` grade described in
+the documentation on [Contexts](Contexts.md#global-components-fluid.resolveRoot-and-fluid.resolveRootSingle).
+
+Having shown the basic operation of the _receiver_ of contextual information, we'll now describe the group of utilities, including `fluid.constructSingle` that we just met, which can be
```
It would be better to lay out the idea that there are these two different types of parts of this API up front (i.e. near the beginning of the document), and then describe them in detail in subsequent sections. It would help the user get a picture of the broader context within which the details fit.

amb26 replied:
This has been done

acheetham commented on src/documents/ContextAwareness.md:
```diff
@@ -0,0 +1,327 @@
+Having shown the basic operation of the _receiver_ of contextual information, we'll now describe the group of utilities, including `fluid.constructSingle` that we just met, which can be
+used by integrators and implementors to coordinate the visibility of context names and distributions from them.
+
+Note that the combined effect of the first two defaults blocks shown in this example can be achieved "all-in-one" by a single call to the dedicated utility `fluid.contextAware.makeAdaptation`. 
```
This mention of 'makeAdaptation' should be a link to the section that describes, and preferably be followed by "described below".

acheetham commented on src/documents/ContextAwareness.md:
```diff
@@ -0,0 +1,327 @@
+
+## Making contexts visible and removing them with `fluid.contextAware.makeChecks` and `fluid.contextAware.forgetChecks`
+
+The `checkRecord` structures described in the table above, by default appeal to the existence of some components holding the boolean value <code>true</code> at an option named <code>value</code> - or
```
The first sentence of this paragraph could use improvement: The phrase "appeal to the existence of" is unclear; the context value could refer to a path other than "option.value" (right?); "which is to be compared by equality" is awkward and indirect. I'm not quite sure what this sentence is trying to say, so I'm not sure what alternative to suggest, but perhaps something like this (if this is actually correct) "The checkRecord structures described above compare examine components in the context, comparing specific options to known values." ?

acheetham commented on src/documents/ContextAwareness.md:
```diff
@@ -0,0 +1,327 @@
+## Making contexts visible and removing them with `fluid.contextAware.makeChecks` and `fluid.contextAware.forgetChecks`
+
+The `checkRecord` structures described in the table above, by default appeal to the existence of some components holding the boolean value <code>true</code> at an option named <code>value</code> - or
+else some other value held at this path which is to be compared by equality. The `ContextAwareness` API includes two helper functions to assist integrators in issuing contexts of this form, and
```
"issuing contexts of this form" – is this correct? Is it actually contexts that are being issued? And what do you mean by "issue"?

acheetham commented on src/documents/ContextAwareness.md:
````diff
@@ -0,0 +1,327 @@
+fluid.contextAware.makeChecks(<checkStructure>);
+```
+
+The `checkStructure` argument is a hash of `contextName` strings to a `checkEntry` record which is documented in the following table:
````
Perhaps replace "which is documented in..." with ", as described in..."

acheetham replied:
Instead of describing the hash in narrative form, it will probably be easier for the reader to parse if you actually provide a code-structured illustration, like
{
    <contextName>: <checkEntry>,
    <contextName>: <checkEntry>,
    ....
}
(This comment also applies to other narrative descriptions of structures above).

acheetham commented on src/documents/ContextAwareness.md:
```diff
@@ -0,0 +1,327 @@
+    <tbody>
+        <tr>
+            <td><code>func</code>/<code>funcName</code></td>
+            <td><code>String</code> <a href="IoCReferences.md">IoC reference</a></td>
```
The member name is given as "func/funcName" and the description says "a function or function name"; should the Type say "A function or a String IoC reference" ?

acheetham commented on src/documents/ContextAwareness.md:
```diff
@@ -0,0 +1,327 @@
+
+## Defining and broadcasting a fresh adaptation in one operation with `fluid.contextAware.makeAdaptation`
+
+A very common use case is to define an adaptation (that is, a `distributeOptions` block which targets the `contextAwareness` area of a collection
```
Is this what an adaptation is? "a distributeOptions block which targets the contextAwareness area of a collection of componentsin the tree" ?

amb26 replied:
This is a kind of thing that can *cause* an adaptation - that is, an entry in a contextAwareness options section which is acted on. Any other way of getting the options there would do as well...

acheetham commented on src/documents/ContextAwareness.md:
```diff
@@ -0,0 +1,327 @@
+
+A very common use case is to define an adaptation (that is, a `distributeOptions` block which targets the `contextAwareness` area of a collection
+of componentsin the tree), and then to create an instance of a single, well-known component which actually broadcasts the adaptation. This was
+what we did in two steps (`fluid.defaults` plus `fluid.constructSingle`) in the above [example `contextAwareness` broadcast](#example-of-dynamically-broadcasting-a-fresh-adaptation)
```
There's something grammatically wrong with this last sentence.

acheetham commented on src/documents/ContextAwareness.md:
````diff
@@ -0,0 +1,327 @@
+fluid.contextAware.makeAdaptation(<adaptationRecord>);
+```
+
+This accepts a single complex structure `adaptationRecord` with a number of fields that need to be filled in:
````
1) No need for the word 'complex' here
2) Perhaps instead of "a number of fields that need to be filled in" maybe "a number of required fields" ?

acheetham commented on src/documents/ChangeApplierAPI.md:
```diff
@@ -234,7 +239,8 @@ applier.modelChanged.addListener(pathSpec, listener, namespace)
 
-**NOTE**: _This style of listening to changes is **not recommended**. Since such a listener can only be attached once a component and its applier have been fully constructed, it will miss observation of the initial transaction as well as any other model changes that have occurred up to the point where it is registered. This implies that the state of the model that it sees cannot be fully predicte ...
+**NOTE**: _This style of listening to changes is **discouraged**, but may be the right choice in some applications_. For example - the listener to be attached may not be available
+at the time the component is constructed. Note that programmatically attached listeners will miss observation of the initial transaction as any other model changes that have occurred up to the point where they are registered.
```
"as any other model changes..." Should this be "as well as any other model changes..." ?

acheetham commented on src/documents/ChangeApplierAPI.md:
```diff
@@ -379,7 +386,7 @@ applier.fireChangeRequest(changeRequest)
 where a **changeRequest** is an object holding the above named parameters in named fields - e.g. `{path: "modelPath", value: "newValue"}` `change` and `fireChangeRequest` reach exactly the same implementation - the only difference is in the packaging of the arguments. For `change` they are spread out as a sequence of 3 arguments, whereas for `fireChangeRequest`, they are packaged up as named fiel ...
 
-The programmatic style for **firing** changes is not strongly discouraged, as the programmatic style for **listening** to changes is - since it does not run into the same lifecycle issues that programmatic listeners do. However, the declarative style for triggering changes should be used wherever it can, and the framework will support more and more schemes for declarative triggering of changes mo ...
+The programmatic style for **firing** changes is less strongly discouraged than the programmatic style for **listening** to changes is - since it does not run into the same lifecycle issues that programmatic listeners do. However, the declarative style for triggering changes should be used wherever it can.
```
Should "firing" be changed to "triggering" here?

acheetham commented on src/documents/ChangeApplierAPI.md:
````diff
@@ -446,29 +453,21 @@ These are not recommended for typical users of the framework, and are not stable
-```
-
-In every case the `options` argument is optional. Where the parameter `model` appears it directly represents the model to be managed by the applier. Where parameter `holder` appears, it represents instead an object which is expected _to be a container for the model_ where the model itself is held in this object at the member named `model`. In framework use this argument will actually hold the own ...
+Instantiating a ChangeApplier manually is no longer recommended in current versions of the framework. Its implementation is tightly bound into its location in 
````
Perhaps instead of "no longer recommended in current versions of the framework" just "is not recommended" ?

acheetham commented on src/documents/ChangeApplierAPI.md:
```diff
@@ -446,29 +453,21 @@ These are not recommended for typical users of the framework, and are not stable
 
-Similarly to manual construction of a ChangeApplier, this is not recommended for normal users of the framework. It may be appropriate for advanced users of the framework who are building higher-level frameworks layered over it.
+A user may be interested in economising on notifications to model updates - by batching these up into a single transaction, there will
+just be a single notification of each listener which is impacted around the model skeleton. This facility is unstable and will be 
```
"This facility is unstable"?? What does that mean? Should I not use it? Up above, you explicitly state when something is not recommended, but you don't do that here. Does that mean it's ok to use? How could it be ok if it's unstable? What might go wrong if I use it?

amb26 replied:
It's ok to use but it is a very annoying and cumbersome API - noone would use it in its current form unless the need were really pressing. Also, it has no what you could call "declarative" or "externalisable" equivalent - that is, it is not possible to describe the required transactional structure from environments outside the process, e.g. by means of the NEXUS. Consult https://wiki.fluidproject.org/display/fluid/New+New+Notes+on+the+ChangeApplier if you are very interested in the details :)
In short, it is both "ok to use for the time being" (since there is no alternative) as well as being an unstable API.

acheetham commented on src/documents/ComponentConfigurationOptions.md:
```diff
@@ -4,30 +4,28 @@ layout: default
+While component developers are free to define whatever options are appropriate for their component, the Infusion Framework supports a number of predefined options.
 
-Some predefined options should not be overridden by integrators: They are strictly for the use of the component developer. This is noted in the descriptions below.
+The particular set of options interpreted by the framework is determined by the [Grades](ComponentGrades.md) that the component is derived from. Developers and integrators
```
The first sentence of this paragraph implies that the framework will only interpret options defined by the specified grades, and not by the current component.

amb26 replied:
I'm not sure that I see that it does - could you explain how it does this, and if possible provide an alternative wording that doesn't?

acheetham commented on src/documents/ComponentConfigurationOptions.md:
```diff
@@ -4,30 +4,28 @@ layout: default
 
-Some predefined options should not be overridden by integrators: They are strictly for the use of the component developer. This is noted in the descriptions below.
+The particular set of options interpreted by the framework is determined by the [Grades](ComponentGrades.md) that the component is derived from. Developers and integrators
+can define further grades which respond to yet further options, which they should document in a similar style if they expect the options to be generally useful. 
```
"document in a similar style"? What do you mean document? Like this current page? similar style to what?

acheetham commented on src/documents/ComponentConfigurationOptions.md:
```diff
@@ -4,30 +4,28 @@ layout: default
-Some predefined options should not be overridden by integrators: They are strictly for the use of the component developer. This is noted in the descriptions below.
+The particular set of options interpreted by the framework is determined by the [Grades](ComponentGrades.md) that the component is derived from. Developers and integrators
+can define further grades which respond to yet further options, which they should document in a similar style if they expect the options to be generally useful. 
+This page briefly describes these predefined options and provides links more information about the related framework functionality.
```
Missing a "to" here: "provides links *to* more information..."

acheetham commented on src/documents/ComponentConfigurationOptions.md:
```diff
@@ -179,80 +195,51 @@ The following options are supported by all component grades:
-  <tr>
-    <th>Notes</th>
-    <td>Some special context names may be available within the subcomponent's definition block, for example <code>{source}</code> and <code>{sourcePath}</code> or <code>{arguments}</code>. This framework facility will be replaced by a more declarative equivalent in time and should be used with caution.</td>
+    <td>A record directing the framework to distribute options from this component to one or more other components in the component tree. Either a single, an <code>Array</code> or <code>Object</code> holding these records is supported.
```
"Either a single, an Array or Object holding these records..." is incomplete (a single what? an array of what? Please clarify.

acheetham commented on src/documents/ComponentConfigurationOptions.md:
```diff
@@ -260,59 +247,62 @@ See also: [Component Grades](ComponentGrades.md)
   <tr>
     <th>Description</th>
-    <td>A <a href="ChangeApplier.md">ChangeApplier</a> object for the model provided with the <code><a href="#model">model</a></code> option.</td>
+    <td>An object containing named definitions of the component's dynamic subcomponents. Rather than exactly one subcomponent being associated with its parent from these records, there may be one subcomponent per element of an array, or one per firing of an event.</td>
```
This reference to 'dynamic subcomponents' should probably link to what documentation we have on this, which is in the "Subcomponent Declaration" page

acheetham commented on src/documents/ComponentConfigurationOptions.md:
```diff
@@ -260,59 +247,62 @@ See also: [Component Grades](ComponentGrades.md)
+
+See also: [Component Grades](ComponentGrades.md)
+
+### `model` ###
```
Why remove the 'see also' link to the changeApplierOptions option in the description of model? The option is still supported...

acheetham commented on src/documents/ComponentConfigurationOptions.md:
```diff
@@ -332,86 +321,76 @@ fluid.defaults("component.name", {
-    },
-    ...
-});</code>
+fluid.defaults("fluid.tests.allChangeRecorder", {
```
This code block is not rendering as code

acheetham commented on src/documents/ComponentConfigurationOptions.md:
```diff
@@ -332,86 +321,76 @@ fluid.defaults("component.name", {
-    },
-    listeners: {
-        onSave: "component.name.saveValidatorFn"
+fluid.defaults("examples.volumeModelRelay", {
```
This code block is not rendering as code

amb26 replied:
It looks the same as all the other examples to me: https://github.com/amb26/infusion-docs/blob/FLUID-5249/src/documents/ComponentConfigurationOptions.md

acheetham commented on src/documents/ComponentConfigurationOptions.md:
```diff
@@ -468,19 +446,51 @@ Component developers are free to define their own additional options.
+  <tr>
+    <th>Example Definition</th>
+    <td><pre>
+fluid.defaults("demo.initGridReorderer", {
```
This code block is not rendering as code.

amb26 replied:
Could you explain what you are seeing here? It appears ok to me: https://github.com/amb26/infusion-docs/blob/FLUID-5249/src/documents/ComponentConfigurationOptions.md

acheetham replied:
Odd. It looks fine now. Sorry about that. Might have been a cache issue?

acheetham commented on src/documents/ComponentConfigurationOptions.md:
```diff
@@ -468,19 +446,51 @@ Component developers are free to define their own additional options.
 
 ## Renderer Components ##
 
 Components defined with a grade of `rendererComponent` are also view components (and hence model components and evented components), so they support
```
Probably remove "evented components" here?

acheetham commented on src/documents/Contexts.md:
```diff
@@ -61,15 +52,106 @@ at a particular component location in the tree are the same, as described in the
-* Finally, context names are sought in the *static environment*, which is a tree of components
-attached globally to the JS VM (or context within it) as a whole.
+* Following this, a search is made in the *resolver root*, which is notionally a separate tree
+of components which holds all components which has been instantiated with the grade `fluid.resolveRoot` anywhere in the
```
This mention of fluid.resolveRoot should be a link to the relevant section and should be followed by "(described below)"

acheetham commented on src/documents/Contexts.md:
```diff
@@ -61,15 +52,106 @@ at a particular component location in the tree are the same, as described in the
+of components which holds all components which has been instantiated with the grade `fluid.resolveRoot` anywhere in the
+current component tree.
 
 ![Diagram showing coloured squared representing components in a component tree](images/IoC-scope.svg)
```
This diagram seems to still illustrate the three kinds of scopes/environments that we used to have, still referencing the dynamic and static environments that are no longer mentioned in this document. The diagram should be udpated to reflect the new reality.

acheetham commented on src/documents/Contexts.md:
```diff
@@ -61,15 +52,106 @@ at a particular component location in the tree are the same, as described in the
+
+All Infusion components are stored in a single-rooted **_component tree_** - this stores every component which
+has been instantiated in a particular **_Infusion context_**. This context is global to the JavaScript context 
+which has loaded Infusion - for example, a frame within a browser, or a node.js process. Even when you 
```
This opening statement is pretty helpful in understanding Contexts. I think it should be placed much closer to the opening paragraph of the page, where "component tree" is first mentioned.

acheetham commented on src/documents/Contexts.md:
```diff
@@ -61,15 +52,106 @@ at a particular component location in the tree are the same, as described in the
+wherever it is constructed in the component tree, will become **_globally visible_** throughout the component
+tree, as if it had been constructed at its root. Secondly, a component which uses the grade 
+`fluid.resolveRootSingle` will be able to arrange that a single instance of it is globally visible at any
+time. Constructing a further instance of such a component will displace the earlier instance from global
```
This explanation of what these two grades do should probably be called out into a table. Buried in a narrative paragraph like this, it's much harder to process.

amb26 replied:
Table added

acheetham commented on src/documents/Contexts.md:
````diff
@@ -61,15 +52,106 @@ at a particular component location in the tree are the same, as described in the
+var that2 = fluid.component({
+    value: "{freeComponent1}.options.freeValue" // error: though global, freeComponent1 is not resolvable
+});
+```
````
We need something much more explicit on this example to indicate that it's an error – more than just a brief comment in the paragraph above it.

acheetham commented on src/documents/Contexts.md:
```diff
@@ -61,15 +52,106 @@ at a particular component location in the tree are the same, as described in the
+JavaScript does not supply schemes for reference tracking, there is no way for the framework to determine when a component is no
+longer being used. The user must call `destroy()` on such instances themselves. 
+
+``javascript
```
This needs three open-quotes; it's not rendering as code without that.

acheetham commented on src/documents/Contexts.md:
````diff
@@ -61,15 +52,106 @@ at a particular component location in the tree are the same, as described in the
+});
+```
+
+However, in the following example, since `rootComponent1` has the grade `fluid.resolveRootSingle`, the instance `root1` can be seen by the
````
It might be helpful to the reader to provide an example of how to fix the above scenario – an example that uses the same component names and only adds the minimum of what would be required to allow that2 to see freeComponent1. This would be in addition to the example below.

amb26 replied:
How would that represent "fixing" the scenario - and, how, specifically, would you like to see it fixed? We could fix it, for example, by never instantiating root2. We could also fix it, for example, by instantiating component that2 before we instantiate root2. We could also fix it by giving that2 a completely different type, as well as root2 a different type, which had a different "singleRootType" field. Do you think showing any of those options might help the reader?

acheetham commented on src/documents/FrameworkConcepts.md:
```diff
@@ -8,31 +8,56 @@ The Fluid Infusion Framework is built on a number of important concepts. This pa
+If everything that is important to you about the object has been successfully cloned, then it is a suitable model object.
 
-Note that this does not prohibit functions, DOM nodes, etc. from appearing in a model, just so long as the meaning of the object tree is not destroyed through the operation of `fluid.copy()`. The point here is that Infusion doesn't impose a formal contract on model objects: as long as `fluid.copy()` succeeds, you're good. For reference, the current implementation of `fluid.copy()` simply forwards ...
+We recommend that all objects stored in the `model` area of a component conform to an even stricter definition - that they are equivalent to their form as serialised as JSON. This implies that they
```
The phrase "as serialized as JSON" seems wrong – probably an extra "as"?

acheetham commented on src/documents/FrameworkConcepts.md:
```diff
@@ -8,31 +8,56 @@ The Fluid Infusion Framework is built on a number of important concepts. This pa
 
-In some parts of the framework, we refer to "EL expressions". This is a somewhat historical phrase that perhaps sounds like it says more than it is trying to. All we mean by EL expressions are dot-separated paths built of names—for example if you had defined an object `zar` with a member `boo` which has a member `baz`, you could access the nested JavaScript property by writing the expression `zar ...
+In some parts of the framework, we refer to "EL expressions". This is a somewhat historical phrase that perhaps sounds like it says more than it is trying to. 
+All we mean by EL expressions are dot-separated paths built from names &emdash; for example if you had defined an object `zar` with a member `boo` which has a member `baz`, 
```
The "&emdash;" is not rendering properly.

amb26 replied:
Oh yes - I had forgotten here that markdown is such a marvellous technology that I need to encode em dashes using the Unicode reference &#8212; ....

amb26 replied:
(that read - ampersand hash 8212 : P )

acheetham commented on src/documents/InfusionEventSystem.md:
```diff
@@ -88,7 +91,15 @@ In general you shouldn't fire any of a component's events unless invited to by i
+Both as part of defaults, and also as supplied instantiation options, an Infusion component can accept a structure named `listeners`. 
+In the simplest form, the keys of the `listeners` structure are taken from the set of `events` present in the component's [Grade](ComponentGrades.md), and the values are either single listener specifications or arrays of listener specifications. 
+A ***listener specification*** can take a number of forms - either being written as a simple String or Function, or as a full JSON object.
+The standard way of declaring a listener using Infusion's [IoC](to-do/IoCInversionOfControl.md) system is to supply the name of a global function using the member **`funcName`** or 
```
These two possibilities would be better presented in a table – more easy to process for the reader.

acheetham commented on src/documents/InfusionEventSystem.md:
```diff
@@ -124,12 +135,63 @@ listeners: {
+    <td><code>args</code> (optional)</td>
+    <td>Any JSON type</td>
+    <td>If supplied, this structure will be used to replace the actual arguments received by the listener, rather than necessarily receiving the exact signature fired by the firer. Any [IoC references](IoCReferences.md) or [expanders](ExpansionOfComponentOptions.md) will be expanded
+    at the time of invoking the listeners, with the additional possibility of referring to the original argument list by using the special context `{arguments}'. Consult [Event Injection and Boiling](EventInjectionAndBoiling.md) for more details of this process.</td>
```
The close quote on {arguments} needs to be a back-quote. All the following text is rendering as code-styled.

acheetham commented on src/documents/InfusionEventSystem.md:
```diff
@@ -124,12 +135,63 @@ listeners: {
+    <td><code>String</code> or <code>Number</code></td>
+    <td>This field allows the configurer to control the sequence in which several listeners to the same event are notified. The recommended form of this field is either `before:otherNamespace` or `after:otherNamespace` where `otherNamespace` represents the `namespace` of some other
+    listener attached to this firer. The framework will sort all the listeners attached to a single firer so that such a listener is fired immediately before or after the other target listener, unless a further constraint positions a third listener in between them.
+    If a group of listeners express a cyclic set of constraints, the framework will signal an error. If there is no listener with the target namespace, the constraint is ignored. There are other possibilities for the `priority` field which are not
```
If we don't recommend the other possibilities, perhaps best to just not mention them at all. The link to the Priorities page should be sufficient.

acheetham commented on src/documents/Invokers.md:
```diff
@@ -53,16 +53,6 @@ If no args are specified, all of the arguments passed into the invoker are sent
             </td>
         </tr>
-        <tr>
-            <td><code>dynamic</code></td>
```
Why was this removed? It's used in examples on this page.

amb26 replied:
"good catch" - removed all remaining references to dynamic

acheetham commented on src/documents/Invokers.md:
```diff
@@ -144,12 +134,8 @@ __Example:__
-        // regular invokers:
         addVal: "xyz.widget.add({that}.staticVal, {arguments}.0)",
-        subtractVal: "{parent}.subtract({arguments}.0, {that}.staticVal)",
-        // dynamic invokers:
```
Are dynamic invokers no longer supported?

amb26 replied:
Yes, and added to list of API changes - I'll push an update once I can figure out how to get the docpad generator to work....

acheetham commented on src/documents/IoCSS.md:
```diff
@@ -4,13 +4,18 @@ layout: default
+means for determining where the target of the distribution is &#8212; is dubbed "IoCSS" because it uses a syntax and idiom very similar to that used in [CSS selectors](https://en.wikipedia.org/wiki/Cascading_Style_Sheets#Selector).
 
-## distributeOptions and IoCSS: Downward-Matching CSS-Like Context Selectors For Options Forwarding ##
+## `distributeOptions and IoCSS`: Downward-Matching CSS-Like Context Selectors For Options Forwarding ##
```
We might want to change the capitalization of this header to "sentence capitalization"

acheetham commented:
General comment relevant to every file: code styling in headers breaks internal links to those headers. Please don't use it. I know it looks better, but it breaks links.

acheetham commented on src/documents/IoCSS.md:
```diff
@@ -4,13 +4,18 @@ layout: default
+when writing powerful authoring tools, this kind of power is completely indispensible.
 
-## Background ##
+## Background for `distributeOptions`##
```
I wonder if most of this background information would be better placed after the technical information that currently follows it. Otherwise, a reader who just needs to look up the formats needs to wade through this information first.

acheetham commented on src/documents/IoCSS.md:
```diff
@@ -47,74 +52,78 @@ fluid.uiOptions(".my-uio-container", {
 ## distributeOptions format ##
 
-The `distributeOptions` option is a top-level block supported by every IoC-configured component, holding an array of objects (or single object) containing the following properties:
+The `distributeOptions` option is a top-level block supported by every IoC-configured component, holding an array of records, hash of records, or single record containing the following properties:
```
Above, you changed "every IoC-configured component" to "every Infusion component" – should that change also be made here?

acheetham commented on src/documents/IoCSS.md:
```diff
@@ -47,74 +52,78 @@ fluid.uiOptions(".my-uio-container", {
+|`source`|(Mutually exclusive with `record`) An [IoC expression](IoCReferences.md) into the options structure of the source component, referencing what to copy to the target.|
+|&nbsp;&nbsp;`removeSource`|(Only possible if `source` is used) true/false: If true, the `source` options block is removed from its original site in the options structure when it is forwarded to the `target`.|
+|&nbsp;&nbsp;`exclusions`|(Only possible if `source` is used) A list of EL paths into the `source` material which should not be forwarded. Whether or not `removeSource` is used, these will be retained in their original position in the source component's options.|
+|`priority`|(Optional) a [Priority specification](Priorities.md) specifying how the priority of this distribution is to be resolved with respect to any other options distributions which are targetted at the same component in the tree as this one. Typically one will write `after:otherNamespace` to indicate that this distribution should be stronger than (should merge options on top of) another dist ...
```
The parenthetical phrase "(should merge options on top of)" would benefit from "i.e." at the beginning.

acheetham commented on src/documents/IoCSS.md:
```diff
@@ -47,74 +52,78 @@ fluid.uiOptions(".my-uio-container", {
-|E#myid|matches any component with a context name of E with id equal to myid (of no use to developers since component ids cannot be predicted)|
+|`*`|matches any component - universal selector|
+|`/`|matches the **global root component** of the entire component tree - may only be used in the head position of an IoCSS selector|
+|`E` or `&E`|matches any component holding a context name of `E` - the `&` character may be omitted for the first context in a group. There is special support for the context `that` as with standard IoC context matching|
```
What does the & character do??

amb26 replied:
It begins a selector segment which matches a context name - can you think of a way I could make this any more clear?

acheetham commented on src/documents/IoCSS.md:
```diff
@@ -47,74 +52,78 @@ fluid.uiOptions(".my-uio-container", {
+|`*`|matches any component - universal selector|
+|`/`|matches the **global root component** of the entire component tree - may only be used in the head position of an IoCSS selector|
+|`E` or `&E`|matches any component holding a context name of `E` - the `&` character may be omitted for the first context in a group. There is special support for the context `that` as with standard IoC context matching|
+|`#myid`|matches the component with id equal to myid (of no use to developers since component ids cannot be predicted)|
```
If this is of no use to developers, should we remove mention of it?

amb26 replied:
It may be of use to framework authors

amb26 commented:
On 14/08/2015 17:11, Anastasia Cheetham wrote:
> General comment relevant to every file: code styling in headers breaks internal links to those headers.
> Please don't use it. I know it looks better, but it breaks links.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/fluid-project/infusion-docs/pull/63#issuecomment-131165179>.
>

Removing code styling won't fix this issue. gh markdown and docpad markdown in rendering anchor ids differ 
in many ways - for example, gh eliminates period characters and docpad doesn't. The code styling issue is 
only relevant when the first word in a title is code-styled, in which case docpad inserts an extra leading 
"-". I'll take care to fix up all our links to conform to the docpad rules.

acheetham commented on src/documents/IoCSS.md:
```diff
@@ -140,3 +149,61 @@ fluid.defaults("fluid.moduleLayoutHandler", {
+
+## Example: broadcast via root to all components ##
+
+([All Husnock, everywhere](http://en.memory-alpha.org/wiki/Husnock)) 
```
I think this is a little bit too obscure to actually contribute anything useful to this page.

acheetham commented on src/documents/ModelRelay.md:
```diff
@@ -4,33 +4,58 @@ layout: default
 
-The Infusion Model Relay system is a powerful scheme for supplying declarative configuration which specifies rules for keeping multiple pieces of model state around the component tree automatically up to date with each other's changes. Because each model relay rule supplies a link between an inner core of models scattered around the component tree, around which other listeners in the periphery of ...
+The Infusion Model Relay system is a powerful scheme for supplying declarative configuration which specifies rules for keeping multiple pieces of model state around the component tree automatically up to date with each other's changes. 
+We sometimes use the term _**model skeleton**_ to refer to a connected set of models around the component tree which are related together by relay rules. This appeals to the idea that these models form an inner core which
```
This opening paragraph seems too long for our convention of a "large-text" first paragraph. Perhaps break it after the first sentence?

acheetham replied:
Actually the rest of this opening paragraph seems a bit too philosophical for this documentation.

amb26 replied:
Do you mean "philosophical" as in "relating or devoted to the study of the fundamental nature of knowledge, reality, and existence"?

colinbdclark replied:
Having read through this section myself, I think it's quite helpful--not too philosophical at all. It provides a means to visualize the way models are related to each other in a component tree.

acheetham commented on src/documents/ModelRelay.md:
```diff
@@ -4,33 +4,58 @@ layout: default
 ---
 
-The Infusion Model Relay system is a powerful scheme for supplying declarative configuration which specifies rules for keeping multiple pieces of model state around the component tree automatically up to date with each other's changes. Because each model relay rule supplies a link between an inner core of models scattered around the component tree, around which other listeners in the periphery of ...
+The Infusion Model Relay system is a powerful scheme for supplying declarative configuration which specifies rules for keeping multiple pieces of model state around the component tree automatically up to date with each other's changes. 
```
I think this should be "that" instead of "which".

acheetham commented on src/documents/ModelRelay.md:
```diff
@@ -4,33 +4,58 @@ layout: default
 ### Two Styles for Configuring Model Relay ###
 
-There are two primary styles of configuration for setting up model relay - firstly, using the _implicit syntax_ which just consists of [IoC References](IoCReferences.md) from the model configuration for one model-bearing component to another - that is, in the component's configuration under the top-level model entry. Secondly, the _explicit syntax_ involves an entry in the component's top-level ` ...
+There are two primary styles of configuration for setting up model relay &#8212; firstly, using the _implicit syntax_ which just consists of [IoC References](IoCReferences.md) from the model configuration for one model-bearing component to another &#8212;
```
These two styles might be better presented as a bulleted or numbered definition list.

acheetham commented on src/documents/ModelRelay.md:
```diff
@@ -4,33 +4,58 @@ layout: default
+
+Every Infusion component descended from the grade `fluid.modelComponent` (a model-bearing component) supports a configuration area named `modelRelay` in which these rules can be defined.
 
 ### Two Styles for Configuring Model Relay ###
```
The headers in this page should be converted to sentence-capitalization style.

acheetham commented on src/documents/ModelRelay.md:
```diff
@@ -93,9 +128,22 @@ This style is used when we require a [Model Transformation](to-do/ModelTransform
         <tr>
             <td>transform</td>
```
'transform' should be code-styled

acheetham commented on src/documents/ModelRelay.md:
```diff
@@ -67,21 +102,21 @@ This style is used when we require a [Model Transformation](to-do/ModelTransform
+            <td><code>target</code> (required)</td>
             <td>String (EL Reference or short local path)</td>
-            <td>The target model path to be linked by this relay rule</td>
+            <td>The target model path to be linked by this relay rule. This option must always be present.</td>
```
Do we need to add "This option must always be present" when we've already added "(required)"?

acheetham commented on src/documents/ModelRelay.md:
```diff
@@ -93,9 +128,22 @@ This style is used when we require a [Model Transformation](to-do/ModelTransform
+        <tr>
+            <td><code>forward</code> (optional)</td>
+            <td>One of the strings <code>never</code>, <code>initOnly</code>, <code>liveOnly</code> or <code>always</code> (the default)</td>
+            <td>Control the situations in which the forward leg of the transform operates - <code>initOnly</code> will have it only operated during the <a href="#the-initial-transaction">initial transaction</a> that this 
```
This should include a link to the "Controlling Propagation Through a Relay Rule" section, which explains forward and backward.

acheetham commented on src/documents/ModelRelay.md:
```diff
@@ -93,9 +128,22 @@ This style is used when we require a [Model Transformation](to-do/ModelTransform
+        <tr>
+            <td><code>backward</code> (optional)</td>
+            <td>One of the strings <code>never</code>, <code>initOnly</code>, <code>liveOnly</code> or <code>always</code> (the default)</td>
+            <td>Control the situations in which the backward leg of the transform operates &#8212; assuming that it is already an invertible transformation. The same meaning for the codes as for <code>forward</code>
```
This should include a link to the "Controlling Propagation Through a Relay Rule" section, which explains forward and backward.

acheetham commented on src/documents/ModelRelay.md:
```diff
@@ -129,10 +180,42 @@ that.child.applier.change("volume", 1);
-In general those transformations which are _**invertible**_ are the best choice for this kind of linkage. If the transforms are not invertible (or have no inverse defined in the framework), the updates will propagate only in one direction. This can still be highly useful.
+In general those transformations which are _**invertible**_ are the best choice for this kind of linkage. If the transforms are not invertible (or have no inverse defined in the framework), 
+the updates will propagate only in one direction. This can still be highly useful. In addition to its invertibility, the propagation of updates through a relay rule can be fine-tuned by the options
+`forward` and `backward`:
```
I don't think we should have a colon at the end of a section like this. Perhaps instead the phrase ", described below" or ", described in the next section".

acheetham commented on src/documents/ModelRelay.md:
```diff
@@ -129,10 +180,42 @@ that.child.applier.change("volume", 1);
+
+A rule of `initOnly` is less often useful, but can be helpful in controlling the process whereby initial default values of more abstract kinds of models are computed.
+
+Compare these directives with the related ones used for source guarding in [model listeners](ChangeApplierAPI.md#source-tracking-and-filtering-in-model-listener-blocks) (`init`, `relay` and `local`).
```
This sentence might be more useful if you actually said something about the comparison.

acheetham commented on src/documents/tutorial-gettingStartedWithInfusion/BasicComponentCreation-Components.md:
```diff
@@ -18,11 +18,11 @@ The definition of a component involves two things:
 ## Grade and Default Options ##
 
-A component's grade and any default options are registered with the framework using a call to [`fluid.defaults`](https://github.com/fluid-project/infusion/blob/infusion-1.5/src/framework/core/js/Fluid.js#L1519-L1539), which has two parameters: the component name and an object containing the defaults. The parent grades for the component are specified in an array in the defaults called `gradeNames` ...
+A component's grade and any default options are registered with the framework using a call to [`fluid.defaults`](https://github.com/fluid-project/infusion/blob/infusion-1.5/src/framework/core/js/Fluid.js#L1519-L1539), which has two parameters: the component name and an object containing the defaults. The parent grades for the component are specified in an array in the defaults called `gradeNames` ...
```
This link is to the infusion-1.5 branch of the code. Can it be updated?

acheetham commented on src/documents/tutorial-gettingStartedWithInfusion/BasicComponentCreation-Components.md:
```diff
@@ -87,7 +87,7 @@ fluid.defaults("fluid.progress", {
 All components have a **creator function**: a public function that is invoked to instantiate the component. In general, the framework will instantiate the creator function for you automatically, given the component's default options. When your component is registered as a [subcomponent](../SubcomponentDeclaration.md) of another, the framework will also take responsibility for calling the creator  ...
 
-* **little**, **evented** and **model** components accept a single argument: options
+* plain, **evented** and **model** components accept a single argument: options
```
We don't have 'evented' components anymore, do we?

acheetham commented on src/documents/tutorial-gettingStartedWithInfusion/EventedComponents.md:
```diff
@@ -9,23 +9,24 @@ Part of the [Getting Started with Infusion Tutorial](GettingStartedWithInfusion.
 ---
 
-Many times, you will be creating a component that works in an environment where other things are operating, and it will probably need to notify those other things of key **events** in its lifecycle. Events can be used to trigger changes in the visual appearance of a component, or actions by other components. For example:
+Many times, you will be creating a component that works in an environment where other things are operating, and it will probably need to notify those other things of key **events** in its lifecycle. 
```
This entire page is no longer appropriate: We don't have a specific "evented" component as one of the types of components to choose from. Any information here that we want to keep in this tutorial should be moved to another page, and the other page that mention evented components and/or link to this page should be updated (the getting started page, the 'pick a component type' page...)

acheetham commented on src/documents/tutorial-gettingStartedWithInfusion/ModelComponents.md:
```diff
@@ -18,15 +18,14 @@ The Infusion Framework provides supports for model-bearing components. When you
 
-* specify a grade of `fluid.modelComponent`, or a grade derived from it (such as `fluid.standardComponent`, `fluid.viewComponent`, etc.) as part of your component's parent grades
-    * <div class="infusion-docs-note"><strong>Note:</strong> For the Infusion 1.5 release cycle, the grade of `fluid.modelRelayComponent` gives access to more modern ChangeApplier facilities, and `fluid.modelComponent` is retained for backwards compatibility with older code. The compatibility implementation will be removed before Infusion 2.0.</div>
+* specify a grade of `fluid.modelComponent`, or a grade derived from it (such as `fluid.modelComponent`, `fluid.viewComponent`, etc.) as part of your component's parent grades
```
Remove "fluid.modelComponent" from the "such as" parenthetical phrase.

amb26 commented:
@colinbdclark , @acheetham  - ready for another look. I've also made some fixes to styling, global structure and the docpad scripts.

acheetham commented on site-structure.json:
```diff
@@ -39,14 +39,25 @@
                 [
                     { "pageName": "Events", "href": "/Events.html" },
                     { "pageName": "Infusion Event System", "href": "/InfusionEventSystem.html" },
+                    { "pageName": "Priorities", "href": "/Priorities.html" },
```
Given that priorities apply to so much more than just events, I wonder if we should put this in a different section of the ToC - maybe the "Defining Components" section, since that's where most of the broadly-relevant stuff is.

acheetham commented on src/layouts/default.html.handlebars:
```diff
@@ -36,11 +38,10 @@
                     <span class="fl-prefsEditor-buttons">
                         <button id="reset" class="flc-prefsEditor-reset fl-prefsEditor-reset"><span class="fl-icon-undo"></span> Reset</button>
-                        <button id="show-hide" class="flc-slidingPanel-toggleButton fl-prefsEditor-showHide"> Show/Hide</button>
+                        <button id="show-hide" class="flc-slidingPanel-toggleButton fl-prefsEditor-showHide">+ Show Display Preferences</button>
```
This change is not necessary: UIO will replace the button text. In fact, leaving it as "show/hide" is a helpful debugging aid, since it allows you to tell at a glance if UIO didn't initialize.

amb26 replied:
This change is necessary to avoid the appalling jank as the UIO initialiser replaces the button text very late into initialisation of the page. Hopefully we can continue to fix the other sources of jank bugs such as https://issues.fluidproject.org/browse/FLUID-5481 since they give our docs site an amateurish appearance. I don't believe that any of our typical docs users will find it valuable to tell "at a glance" whether UIO initialized or not : P

acheetham commented:
The file Events.md doesn't really seem to serve much purpose. Could we get rid of it?

acheetham commented:
There are some problems with the UIO integration. It works and all, but a) the button is about one pixel high; b) when the panel opens and the page slides down, the ToC doesn’t also slide down.

acheetham commented:
I like the separate scrolling for the ToC and the content, but on pages that are very short, the footer is floating  up in the middle of the page. We should try to have the minimum height of the page be the window height.

acheetham commented:
There's supposed to be a gap between the main header of the page and the upper black nav bar, but that seems to be gone now, and it looks rather crowded.

acheetham commented:
The ToC link to APIChangesFrom1_5to2_0.html doesn't actually contain any text; can't tell why. The link is there in the generated html, but without any text in it.

acheetham commented on src/documents/APIChangesFrom1_5To2_0.md:
```diff
@@ -8,20 +8,54 @@ This page contains a list of the features, APIs, and etc. that have changed in I
+### Core Framework Changes ###
 
-Rename "fluid.prefs.enactors" to "fluid.prefs.enactor"
+This section describes major APIs that were in common use. For information about less widely-used features removed in 2.0, consult [Deprecations in 1.5)(DeprecationsIn1_5.md).
```
Typo in this link: ) should be ]

amb26 commented:
Addressed most of the recent comments - UIO is working again, but I didn't have much luck with the footer - I don't think this can be tackled without changing the containment structure of the markup significantly (I already moved up the footer several levels in the document to prevent it bashing into the "position:fixed" TOC at the left). Perhaps @jhung has some ideas - Cheers

acheetham commented on src/documents/tutorial-gettingStartedWithInfusion/PickAComponentType.md:
```diff
@@ -13,16 +13,15 @@ The Infusion Framework provides lots of supports for creating components. We hav
+* **view components** support models, and also add support for views.
 * **renderer components** are **view components** with the [Renderer](../Renderer.md) added.
 
 ![A venn diagram showing the composition of grades](../images/component-grades-venn-diagram.svg)
```
This venn diagram needs updating.

acheetham commented on src/documents/tutorial-gettingStartedWithInfusion/BasicComponentCreation-Components.md:
```diff
@@ -0,0 +1,271 @@
+
+## Grade and Default Options ##
+
+A component's grade and any default options are registered with the framework using a call to [`fluid.defaults`](https://github.com/fluid-project/infusion/blob/infusion-1.5/src/framework/core/js/Fluid.js#L1519-L1539), which has two parameters: the component name and an object containing the defaults. The parent grades for the component are specified in an array in the defaults called `gradeNames` ...
```
This paragraph refers to "little component" – we're not using that phrase anymore, are we?

acheetham replied:
The link to the code for "fluid.defaults" is still pointing to the 1.5 tag.

acheetham commented:
The link at the bottom of the "getting started" tutorial's Model Components page still points to the now non-existent Evented Components page

acheetham commented:
The paragraph at the top of the "getting started" tutorial's View Components page still refers to little and evented components.

acheetham commented:
The "Defining components" link at the top of the "Using String Templates" tutorial still links to the little components page

acheetham commented on src/documents/ComponentConfigurationOptions.md:
```diff
@@ -112,40 +110,63 @@ The following options are supported by all component grades:
+    console.log("Event listener received number " + number + " and condition " + condition);
+};
+
+fluid.defaults("examples.eventedComponent", {
```
Does it make sense to use "evented" in the name of the component, now that everything is evented by default?

acheetham commented:
The "getting started" tutorial's Renderer Components page still has a link to the Evented Components page

acheetham commented:
The Event Injection and Boiling page has a link with link text "eventedComponent" – should probably just be "comonent"?

acheetham commented on src/documents/FrameworkConcepts.md:
```diff
@@ -8,31 +8,54 @@ The Fluid Infusion Framework is built on a number of important concepts. This pa
+Events in Infusion are model-oriented, and aren't specific to the DOM. [Events](InfusionEventSystem.md) have a very plain implementation in Infusion &#8212; an event here is really just another kind of function call. 
+Any function signature can be an event signature, any function can be an event listener, and an event's `fire` method is a plain function handle that can be handed around just like any other function. 
+There is no special kind of "Event Object" that gets handed around to event listeners, and anyone can easily define a new event by simply calling `fluid.makeEventFirer()`. 
+In practice, users should use the events which are created automatically by the framework as part of the initialisation of every [Component](tutorial-gettingStartedWithInfusion/EventedComponents.md). 
```
This link is still referencing the EventedComponents page

acheetham commented on src/documents/InfusionEventSystem.md:
````diff
@@ -46,11 +46,13 @@ A top-level options block named **`events`** is supported on every component der
 For every such entry in the `events` section of a component's options, the framework will construct a corresponding ***event firer*** with the same name in the `events` section of the constructed component. The most common use of an event firer is to call its member named `fire` with some set of arguments. Here is a simple, self-contained example:
 
 ```javascript
 fluid.defaults("examples.eventedComponent", {
````
Does it make sense to keep using "evented" in the example component names?

acheetham commented on src/documents/ModelRelay.md:
```diff
@@ -129,10 +181,42 @@ that.child.applier.change("volume", 1);
+the updates will propagate only in one direction. This can still be highly useful. In addition to its invertibility, the propagation of updates through a relay rule can be fine-tuned by the options
+`forward` and `backward`, as described in the following section.
+
+####Controlling propagation through a relay rule####
```
These headers don't seem to be rendering as headers.

acheetham commented:
The link to the "API Changes from 1.5 to 2.0" page is still broken in the ToC

amb26 commented:
"API Changes from 1.5 to 2.0" works for me - what do you observe when operating the link? Does the URL look correct? Does a file like the one it is pointing to exist in the "out" directory?

amb26 commented:
Ready for further further review
